File: split/macrocall.pas
procedure macrocall;
label
  10, 22, 30, 31, 40;
var
  r: halfword;
  p: halfword;
  q: halfword;
  s: halfword;
  t: halfword;
  u, v: halfword;
  rbraceptr: halfword;
  n: smallnumber;
  unbalance: halfword;
  m: halfword;
  refcount: halfword;
  savescannerstatus: smallnumber;
  savewarningindex: halfword;
  matchchr: ASCIIcode;
begin
  savescannerstatus := scannerstatus;
  savewarningindex := warningindex;
  warningindex := curcs;
  refcount := curchr;
  r := mem[refcount].hh.rh;
  n := 0;
  if eqtb[5293].int > 0 then{401:}

  begin
    begindiagnostic;
    println;
    printcs(warningindex);
    tokenshow(refcount);
    enddiagnostic(false);
  end{:401}
;
  if mem[r].hh.lh <> 3584 then{391:}

  begin
    scannerstatus := 3;
    unbalance := 0;
    longstate := eqtb[curcs].hh.b0;
    if longstate >= 113 then
      longstate := longstate - 2;
    repeat
      mem[29997].hh.rh := 0;
      if (mem[r].hh.lh > 3583) or (mem[r].hh.lh < 3328) then
        s := 0
      else
      begin
        matchchr := mem[r].hh.lh - 3328;
        s := mem[r].hh.rh;
        r := s;
        p := 29997;
        m := 0;
      end;
      {392:}
22:
        gettoken;
      if curtok = mem[r].hh.lh then{394:}

      begin
        r := mem[r].hh.rh;
        if (mem[r].hh.lh >= 3328) and (mem[r].hh.lh <= 3584) then
        begin
          if curtok < 512 then
            alignstate := alignstate - 1;
          goto 40;
        end
        else
          goto 22;
      end{:394}
;
      {397:}
if s <> r then
        if s = 0 then{398:}

        begin
          begin
            if interaction = 3 then;
            printnl(262);
            print(650);
          end;
          sprintcs(warningindex);
          print(651);
          begin
            helpptr := 4;
            helpline[3] := 652;
            helpline[2] := 653;
            helpline[1] := 654;
            helpline[0] := 655;
          end;
          error;
          goto 10;
        end{:398}

        else
        begin
          t := s;
          repeat
            begin
              q := getavail;
              mem[p].hh.rh := q;
              mem[q].hh.lh := mem[t].hh.lh;
              p := q;
            end;
            m := m + 1;
            u := mem[t].hh.rh;
            v := s;
            while true do
            begin
              if u = r then
                if curtok <> mem[v].hh.lh then
                  goto 30
                else
                begin
                  r := mem[v].hh.rh;
                  goto 22;
                end;
              if mem[u].hh.lh <> mem[v].hh.lh then
                goto 30;
              u := mem[u].hh.rh;
              v := mem[v].hh.rh;
            end;
            30:
              t := mem[t].hh.rh;
          until t = r;
          r := s;
        end{:397}
;
      if curtok = partoken then
        if longstate <> 112 then{396:}

        begin
          if longstate = 111 then
          begin
            runaway;
            begin
              if interaction = 3 then;
              printnl(262);
              print(645);
            end;
            sprintcs(warningindex);
            print(646);
            begin
              helpptr := 3;
              helpline[2] := 647;
              helpline[1] := 648;
              helpline[0] := 649;
            end;
            backerror;
          end;
          pstack[n] := mem[29997].hh.rh;
          alignstate := alignstate - unbalance;
          for m := 0 to n do
            flushlist(pstack[m]);
          goto 10;
        end{:396}
;
      if curtok < 768 then
        if curtok < 512 then{399:}

        begin
          unbalance := 1;
          while true do
          begin
            begin
              begin
                q := avail;
                if q = 0 then
                  q := getavail
                else
                begin
                  avail := mem[q].hh.rh;
                  mem[q].hh.rh := 0;
                  {dynused:=dynused+1;}
end;
              end;
              mem[p].hh.rh := q;
              mem[q].hh.lh := curtok;
              p := q;
            end;
            gettoken;
            if curtok = partoken then
              if longstate <> 112 then{396:}

              begin
                if longstate = 111 then
                begin
                  runaway;
                  begin
                    if interaction = 3 then;
                    printnl(262);
                    print(645);
                  end;
                  sprintcs(warningindex);
                  print(646);
                  begin
                    helpptr := 3;
                    helpline[2] := 647;
                    helpline[1] := 648;
                    helpline[0] := 649;
                  end;
                  backerror;
                end;
                pstack[n] := mem[29997].hh.rh;
                alignstate := alignstate - unbalance;
                for m := 0 to n do
                  flushlist(pstack[m]);
                goto 10;
              end{:396}
;
            if curtok < 768 then
              if curtok < 512 then
                unbalance := unbalance + 1
              else
              begin
                unbalance := unbalance - 1;
                if unbalance = 0 then
                  goto 31;
              end;
          end;
          31:
            rbraceptr := p;
          begin
            q := getavail;
            mem[p].hh.rh := q;
            mem[q].hh.lh := curtok;
            p := q;
          end;
        end{:399}

        else{395:}

        begin
          backinput;
          begin
            if interaction = 3 then;
            printnl(262);
            print(637);
          end;
          sprintcs(warningindex);
          print(638);
          begin
            helpptr := 6;
            helpline[5] := 639;
            helpline[4] := 640;
            helpline[3] := 641;
            helpline[2] := 642;
            helpline[1] := 643;
            helpline[0] := 644;
          end;
          alignstate := alignstate + 1;
          longstate := 111;
          curtok := partoken;
          inserror;
          goto 22;
        end{:395}

      else{393:}

      begin
        if curtok = 2592 then
          if mem[r].hh.lh <= 3584 then
            if mem[r].hh.lh >= 3328 then
              goto 22;
        begin
          q := getavail;
          mem[p].hh.rh := q;
          mem[q].hh.lh := curtok;
          p := q;
        end;
      end{:393}
;
      m := m + 1;
      if mem[r].hh.lh > 3584 then
        goto 22;
      if mem[r].hh.lh < 3328 then
        goto 22;
      40:
        if s <> 0 then{400:}

        begin
          if (m = 1) and (mem[p].hh.lh < 768) and (p <> 29997) then
          begin
            mem[rbraceptr].hh.rh := 0;
            begin
              mem[p].hh.rh := avail;
              avail := p;
              {dynused:=dynused-1;}
end;
            p := mem[29997].hh.rh;
            pstack[n] := mem[p].hh.rh;
            begin
              mem[p].hh.rh := avail;
              avail := p;{dynused:=dynused-1;}

            end;
          end
          else
            pstack[n] := mem[29997].hh.rh;
          n := n + 1;
          if eqtb[5293].int > 0 then
          begin
            begindiagnostic;
            printnl(matchchr);
            printint(n);
            print(656);
            showtokenlist(pstack[n - 1], 0, 1000);
            enddiagnostic(false);
          end;
        end{:400}
{:392}
;
    until mem[r].hh.lh = 3584;
  end{:391}
;
  {390:}
while (curinput.statefield = 0) and (curinput.locfield = 0) and (curinput.indexfield <> 2) do
    endtokenlist;
  begintokenlist(refcount, 5);
  curinput.namefield := warningindex;
  curinput.locfield := mem[r].hh.rh;
  if n > 0 then
  begin
    if paramptr + n > maxparamstack then
    begin
      maxparamstack :=
        paramptr + n;
      if maxparamstack > paramsize then
        overflow(636, paramsize);
    end;
    for m := 0 to n - 1 do
      paramstack[paramptr + m] := pstack[m];
    paramptr := paramptr + n;
  end{:390}
;
  10:
    scannerstatus := savescannerstatus;
  warningindex := savewarningindex;
end;

File: cpp/src/macrocall.cpp
#include "macrocall.h"
#include "impression.h"
#include "equivalent.h"
#include "noeud.h"
#include "runaway.h"
#include "erreur.h"
#include "lecture.h"
#include "getnext.h"
#include "alignement.h"
#include "etat.h"
#include "initprim.h"

static TokenList pstack[9];

void macrocall(Token t)
{
	auto savewarningindex = warningindex;
	warningindex = t.cs;
	TokenList refcount;
	refcount.token_ref_count = t.chr;
	halfword r = 0;
	smallnumber n = 0;
	if (tracing_macros() > 0)
		diagnostic("\n"+cs(warningindex)+tokenshow(&refcount));
	ASCIIcode matchchr;
	if (refcount.list.back() != end_match_token) 
	{
		halfword unbalance = 0;
		longstate = t.cs >= hash_base ? eqtb_cs[t.cs-hash_base].type : eqtb_active[t.cs-active_base].type;
		if (longstate >= outer_call)
			longstate -= 2;
		do
		{
			halfword s;
			halfword m;
			if (refcount.list[r] > match_token+255 || refcount.list[r] < match_token)
				s = 0;
			else
			{
				matchchr = refcount.list.back()-match_token;
				s = ++r;
				m = 0;
			}
			while (true)
			{
				t = gettoken(matching);
				if (t.tok == refcount.list[r])
				{
					r++;
					if (refcount.list[r] >= match_token && refcount.list[r] <= end_match_token)
					{
						if (t.tok < left_brace_limit) 
							alignstate--;
						break;
					}
					else
						continue;
				}
				if (s != r)
					if (s == 0)
					{
						error("Use of "+scs(warningindex)+" doesn't match its definition", "If you say, e.g., `\\def\\a1{...}', then you must always\nput `1' after `\\a', since control sequence names are\nmade up of letters only. The macro here has not been\nfollowed by the required stuff, so I'm ignoring it.");
						warningindex = savewarningindex;
						return;
					}
					else
					{
						auto tt = s;
						bool l22 = false;
						do
						{
							tempHead.list.push_back(refcount.list[tt]);
							m++;
							
							for (halfword u = tt+1, v = s; true; u++, v++)
							{
								if (u == r)
									if (t.tok != refcount.list[v])
										break;
									else
									{
										r = v+1;
										l22 = true;
										break;
									}
								if (refcount.list[u] != refcount.list[v])
									break;
							}
							if (!l22)
								tt++;
						} while (tt != r && !l22);
						if (l22)
							continue;
						r = s;
					}
				if (t.tok == partoken)
					if (longstate != long_call)
					{
						if (longstate == call)
						{
							runaway(matching);
							backerror(t, "Paragraph ended before "+scs(warningindex)+" was complete", "I suspect you've forgotten a `}', causing me to apply this\ncontrol sequence to too much text. How can we recover?\nMy plan is to forget the whole thing and hope for the best.");
						}
						pstack[n].list = tempHead.list;
						alignstate -= unbalance;
						for (int m = 0; m <= n; m++)
							pstack[m].list.clear();
						warningindex = savewarningindex;
						return;
					}
				if (t.tok < right_brace_limit)
					if (t.tok < left_brace_limit)
					{
						unbalance = 1;
						while (unbalance)
						{
							tempHead.list.push_back(t.tok);
							t = gettoken(matching);
							if (t.tok == partoken)
								if (longstate != long_call)
								{
									if (longstate == call)
									{
										runaway(matching);
										backerror(t, "Paragraph ended before "+scs(warningindex)+" was complete", "I suspect you've forgotten a `}', causing me to apply this\ncontrol sequence to too much text. How can we recover?\nMy plan is to forget the whole thing and hope for the best.");
									}
									pstack[n].list = tempHead.list;
									alignstate -= unbalance;
									for (int m = 0; m <= n; m++)
										pstack[m].list.clear();
									warningindex = savewarningindex;
									return;
								}
							if (t.tok < right_brace_limit)
								if (t.tok < left_brace_limit)
									unbalance++;
								else
									unbalance--;
						}
						tempHead.list.push_back(t.tok);
					}
					else
					{
						backinput(t);
						inserror(t, "Argument of "+scs(warningindex)+" has an extra }", "I've run across a `}' that doesn't seem to match anything.\nFor example, `\\def\\a#1{...}' and `\\a}' would produce\nthis error. If you simply proceed now, the `\\par' that\nI've just inserted will cause me to report a runaway\nargument that might be the root of the problem. But if\nyour `}' was spurious, just type `2' and it will go away.");
						alignstate++;
						longstate = call;
						t.tok = partoken;
						continue;
					}
				else
				{
					if (t.tok == space_token && refcount.list[r] <= end_match_token && refcount.list[r] >= match_token)
						continue;
					tempHead.list.push_back(t.tok);
				}
				m++;
				if (refcount.list[r] > end_match_token || refcount.list[r] < match_token)
					continue;
				break;
			}
			if (s)
			{
				if (m == 1 && tempHead.list.size() && tempHead.list.back() < right_brace_limit )
				{
					pstack[n].list = tempHead.list;
					pstack[n].list.erase(pstack[n].list.begin());
				}
				else
					pstack[n].list = tempHead.list;
				n++;
				if (tracing_macros() > 0)
					diagnostic(std::string(1, matchchr)+"\n"+std::to_string(n)+"<-"+tokenlist(&pstack[n-1], 1000));
			}
		} while (refcount.list[r] != end_match_token); 
	}
	while (state == 0 && loc == 0 && index != 2)
		endtokenlist();
	beginTokenListMacro(&refcount);
	name = warningindex;
	Loc = r+1;
	if (n > 0)
	{
		if (paramstack.size()+n > paramsize)
			overflow("parameter stack size", paramsize);
		for (int m = 0; m < n; m++)
			paramstack.push_back(pstack[m]);
	}
	warningindex = savewarningindex;
}
